server/block/vine.go

```diff
@@ -0,0 +1,330 @@
+package block
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/block/model"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/go-gl/mathgl/mgl64"
+	"math/rand"
+)
+
+// Vines are climbable non-solid vegetation blocks that grow on walls.
+type Vines struct {
+	replaceable
+	transparent
+	empty
+	sourceWaterDisplacer
+
+	// NorthDirection is true if the vines are attached towards north.
+	NorthDirection bool
+	// EastDirection is true if the vines are attached towards east.
+	EastDirection bool
+	// SouthDirection is true if the vines are attached towards south.
+	SouthDirection bool
+	// WestDirection is true if the vines are attached towards west.
+	WestDirection bool
+}
+
+// CompostChance ...
+func (Vines) CompostChance() float64 {
+	return 0.5
+}
+
+// SideClosed ...
+func (Vines) SideClosed(cube.Pos, cube.Pos, *world.Tx) bool {
+	return false
+}
+
+// HasLiquidDrops ...
+func (Vines) HasLiquidDrops() bool {
+	return false
+}
+
+// FlammabilityInfo ...
+func (Vines) FlammabilityInfo() FlammabilityInfo {
+	return newFlammabilityInfo(15, 100, true)
+}
+
+// BreakInfo ...
+func (v Vines) BreakInfo() BreakInfo {
+	return newBreakInfo(0.2, alwaysHarvestable, func(t item.Tool) bool {
+		return t.ToolType() == item.TypeShears || t.ToolType() == item.TypeAxe
+	}, func(t item.Tool, enchantments []item.Enchantment) []item.Stack {
+		if t.ToolType() == item.TypeShears {
+			return []item.Stack{item.NewStack(v, 1)}
+		}
+		return nil
+	})
+}
+
+// EntityInside ...
+func (Vines) EntityInside(_ cube.Pos, _ *world.Tx, e world.Entity) {
+	if fallEntity, ok := e.(fallDistanceEntity); ok {
+		fallEntity.ResetFallDistance()
+	}
+}
+
+// SetAttachment sets an attachment on the given cube.Direction.
+func (v Vines) SetAttachment(direction cube.Direction, attached bool) Vines {
```

Comment:
Would name this `WithAttachment` to make it clear that this does not modify the original block.